File: pwa-gamespace/test/progress-estimator.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { ProgressEstimator } from "../src/progress-estimator.js";

function createEstimator() {
  const estimator = new ProgressEstimator();
  estimator.reset(0);
  return estimator;
}

test("refreshes speed and ETA once per second, completing immediately", () => {
  const estimator = createEstimator();
  const early = estimator.update(100, 1000, 250);
  assert.equal(early.speedPerSecond, null);
  assert.equal(early.updated, false);

  const middle = estimator.update(500, 1000, 1000);
  assert.equal(middle.speedPerSecond, 500);
  assert.equal(middle.remainingMs, 1000);
  assert.equal(middle.updated, true);

  const burst = estimator.update(900, 1000, 1250);
  assert.equal(burst.speedPerSecond, 500);
  assert.equal(burst.remainingMs, 1000);
  assert.equal(burst.updated, false);
  assert.equal(burst.processed, 900);
  const complete = estimator.update(1000, 1000, 1300);
  assert.equal(complete.remainingMs, 0);
  assert.equal(complete.updated, true);
});

test("duplicate events do not inflate the next rate", () => {
  const estimator = createEstimator();
  estimator.update(1000, 10000, 1000);
  for (let at = 1001; at < 2000; at++) estimator.update(1000, 10000, at);
  const next = estimator.update(2000, 10000, 2000);
  assert.equal(next.speedPerSecond, 1000);
  assert.equal(next.remainingMs, 8000);
});

test("the window forgets old rates after fifteen seconds", () => {
  const estimator = createEstimator();
  for (let second = 1; second <= 20; second++) estimator.update(second * 1000, 100000, second * 1000);
  const changed = estimator.update(22000, 100000, 21000);
  assert.ok(Math.abs(changed.speedPerSecond - 16000 / 15) < 0.001);
  for (let second = 22; second < 35; second++) estimator.update(20000 + (second - 20) * 2000, 100000, second * 1000);
  const settled = estimator.update(50000, 100000, 35000);
  assert.equal(settled.speedPerSecond, 2000);
  assert.equal(settled.remainingMs, 25000);
});

test("irregular events interpolate the window boundary by elapsed time", () => {
  const estimator = createEstimator();
  estimator.update(2000, 100000, 2000);
  estimator.update(6000, 100000, 4000);
  const estimate = estimator.update(36000, 100000, 19000);
  assert.equal(estimate.speedPerSecond, 2000);
  const shifted = estimator.update(46000, 100000, 21000);
  // At 6 s the interpolated count is 10000, giving 36000 bytes in 15 s.
  assert.equal(shifted.speedPerSecond, 2400);
});

test("a pause decays to zero with unknown ETA and can resume", () => {
  const estimator = createEstimator();
  estimator.update(1000, 10000, 1000);
  for (let second = 2; second < 16; second++) estimator.update(1000, 10000, second * 1000);
  const paused = estimator.update(1000, 10000, 16000);
  assert.equal(paused.speedPerSecond, 0);
  assert.equal(paused.remainingMs, null);
  const resumed = estimator.update(2500, 10000, 17000);
  assert.equal(resumed.speedPerSecond, 100);
  assert.equal(resumed.remainingMs, 75000);
});

test("restarts the baseline when counters or time go backwards", () => {
  const estimator = createEstimator();
  estimator.update(800, 1000, 1000);
  const restarted = estimator.update(1, 10, 1200);
  assert.equal(restarted.speedPerSecond, null);
  assert.equal(restarted.remainingMs, null);
  assert.equal(restarted.updated, true);
  assert.equal(estimator.update(6, 10, 2200).speedPerSecond, 5);
  assert.equal(estimator.update(7, 10, 2100).speedPerSecond, null);
  assert.equal(estimator.update(9, 10, 3100).speedPerSecond, 2);
});

test("handles unknown totals and completion before the first estimate", () => {
  const estimator = createEstimator();
  assert.equal(estimator.update(100, 0, 1000).remainingMs, null);
  estimator.reset(0);
  const complete = estimator.update(10, 10, 100);
  assert.equal(complete.speedPerSecond, null);
  assert.equal(complete.remainingMs, 0);
  assert.equal(complete.updated, true);
});

test("large archives retain bounded samples during frequent progress events", () => {
  const estimator = createEstimator();
  const total = 40 * 1024 ** 3;
  let estimate;
  for (let at = 1; at <= 60000; at++) estimate = estimator.update(at * 1024 ** 2 / 10, total, at);
  assert.equal(estimate.speedPerSecond, 100 * 1024 ** 2);
  assert.equal(estimate.remainingMs, 349600);
  assert.ok(estimator.samples.length <= 17);
});

File: apk-gamespace/tests/ProgressEstimatorTest.java
package ru.local.gamespace.loader;

/** Deterministic burst, stall and large-archive scenarios without Android. */
public final class ProgressEstimatorTest {
    private static int checks;

    public static void main(String[] args) {
        ProgressEstimator estimator = new ProgressEstimator(0L);
        close(estimator.update(100L, 1000L, 250L).speedPerSecond, -1.0, "Initial speed is unknown");
        ProgressEstimator.Estimate middle = estimator.update(500L, 1000L, 1000L);
        close(middle.speedPerSecond, 500.0, "First second rate");
        check(middle.remainingMs == 1000L, "First ETA");
        ProgressEstimator.Estimate burst = estimator.update(900L, 1000L, 1250L);
        close(burst.speedPerSecond, 500.0, "Rate held between refreshes");
        check(burst.remainingMs == 1000L, "ETA held between refreshes");
        check(estimator.update(1000L, 1000L, 1300L).remainingMs == 0L, "Completion is immediate");

        estimator = new ProgressEstimator(0L);
        estimator.update(1000L, 10000L, 1000L);
        for (long at = 1001L; at < 2000L; at++) estimator.update(1000L, 10000L, at);
        close(estimator.update(2000L, 10000L, 2000L).speedPerSecond, 1000.0, "Duplicate events do not inflate rate");

        estimator = new ProgressEstimator(0L);
        for (long second = 1L; second <= 20L; second++) estimator.update(second * 1000L, 100000L, second * 1000L);
        close(estimator.update(22000L, 100000L, 21000L).speedPerSecond, 16000.0 / 15.0, "Rate changes gradually");
        for (long second = 22L; second < 35L; second++) estimator.update(20000L + (second - 20L) * 2000L, 100000L, second * 1000L);
        ProgressEstimator.Estimate settled = estimator.update(50000L, 100000L, 35000L);
        close(settled.speedPerSecond, 2000.0, "Old rate leaves the 15-second window");
        check(settled.remainingMs == 25000L, "ETA follows the recent rate");

        estimator = new ProgressEstimator(0L);
        estimator.update(2000L, 100000L, 2000L);
        estimator.update(6000L, 100000L, 4000L);
        close(estimator.update(36000L, 100000L, 19000L).speedPerSecond, 2000.0, "Sparse events use elapsed time");
        close(estimator.update(46000L, 100000L, 21000L).speedPerSecond, 2400.0, "Window boundary is interpolated");

        estimator = new ProgressEstimator(0L);
        estimator.update(1000L, 10000L, 1000L);
        for (long second = 2L; second < 16L; second++) estimator.update(1000L, 10000L, second * 1000L);
        ProgressEstimator.Estimate paused = estimator.update(1000L, 10000L, 16000L);
        close(paused.speedPerSecond, 0.0, "Long pause reaches zero rate");
        check(paused.remainingMs == -1L, "Paused ETA is unknown");
        close(estimator.update(2500L, 10000L, 17000L).speedPerSecond, 100.0, "Resume includes the pause");

        estimator = new ProgressEstimator(0L);
        estimator.update(800L, 1000L, 1000L);
        close(estimator.update(1L, 10L, 1200L).speedPerSecond, -1.0, "Counter restart clears rate");
        close(estimator.update(6L, 10L, 2200L).speedPerSecond, 5.0, "Restart uses a new baseline");
        close(estimator.update(7L, 10L, 2100L).speedPerSecond, -1.0, "Clock reversal clears rate");
        close(estimator.update(9L, 10L, 3100L).speedPerSecond, 2.0, "Clock reversal uses a new baseline");

        estimator = new ProgressEstimator(0L);
        check(estimator.update(100L, 0L, 1000L).remainingMs == -1L, "Unknown total has no ETA");
        estimator = new ProgressEstimator(0L);
        ProgressEstimator.Estimate shortOperation = estimator.update(10L, 10L, 100L);
        check(shortOperation.remainingMs == 0L && shortOperation.speedPerSecond < 0.0, "Subsecond completion has zero ETA");

        estimator = new ProgressEstimator(0L);
        long total = 40L * 1024L * 1024L * 1024L;
        ProgressEstimator.Estimate large = null;
        for (long at = 1L; at <= 60000L; at++) large = estimator.update(at * 1024L * 1024L / 10L, total, at);
        close(large.speedPerSecond, 100.0 * 1024 * 1024, "Rate supports archives above 4 GB");
        check(large.remainingMs == 349600L, "Large archive ETA");
        check(estimator.sampleCount() <= 17, "Frequent events retain bounded samples");
        System.out.println("Progress estimator: " + checks + " checks passed.");
    }

    private static void close(double actual, double expected, String message) {
        check(Math.abs(actual - expected) < 0.001, message + ": " + actual);
    }

    private static void check(boolean condition, String message) {
        if (!condition) throw new AssertionError(message);
        checks++;
    }
}
